Predict

jmercouris opened the pull request:
# Description

This library adds the ability to predict commands for the user. It utilizes the new composite-sequence functionality built into the analysis library. This library is able to scan multiple sequences of varying lengths and make predictions.

# Discussion

There is one problem here. By using a sliding window of one element, we are forced into the following pattern.

Run `execute-command`.
Run `select-next-source` to view the prediction.

Now, the model has been trained, what follows `execute-command` is always `select-next-source`. This will always be your prediction. I did validate that the model still works, however, this prediction is hardly useful. Suggested remedy:

+ Bind the prediction to a keybinding
OR
+ Add a list of commands which are ignored by the model

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [X] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [ ] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [ ] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [ ] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [X] I have updated the existing documentation to match my changes.
  - [ ] I have commented my code in hard-to-understand areas.
  - [X] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [X] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

jmercouris commented:
P.S. yay!

Ambrevar commented on source/changelog.lisp:
```diff
@@ -537,4 +537,8 @@ the selection, and the string to paste instead of the clipboard (respectively)."
+   (:li "New prediction capabilities. Nyxt can now predict your next command, it will
+show up automatically in the execute-command menu. Nyxt uses a stochastic model
+to generate predictions for what your next command will be. The model is stored
+locally and is cleared after every session.")))
```
I think you mean "kept in memory", since it's not stored to disk.

By the way, why not store it to disk?

jmercouris replied:
Sure, it is kept in memory. I wanted to emphasize to the user that it does not rely on ML coming from some server! I'll clarify the wording.

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+;;
+;; We would record the following sequences/outcomes.
+;;
+;; Chain length of 1:
```
This is a bit confusing because the chain length is 2 here.
So either write "chain length of 2" or "chain with 1 transition".

jmercouris replied:
The chain length is in fact one. Because there is one value -> one outcome. I can put a delimeter to make it more obvious.

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+  (mapcar #'element (alexandria:hash-table-values (edges node))))
+
+(defmethod increment ((node node))
+  (incf (occurrences node)))
```
Useless method?

jmercouris replied:
I don't believe so. I don't wish to expose the implementation details even within the program itself.

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+
+(defmethod add-record ((model sequence-model) sequence)
+  (let* ((list-but-last-element (butlast sequence))
+         (last-element (car (last sequence)))
```
Style: Prefer `first` over `car`.

jmercouris replied:
I do usually prefer first, not sure why I used car here. I can't remember my line of reasoning.

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+         (last-element (car (last sequence)))
+         (leaf (alexandria:ensure-gethash list-but-last-element
+                                          (edges model)
+                                          (make-instance 'node))))
```
Instead of using `butlast` and `last` separately, you can use `(sera:halves sequence -1)` which traverses the list only once.

jmercouris replied:
Clever. Done!

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+    (increment (add-edge leaf (make-instance 'element-node :element last-element)))))
+
+(defmethod add-record-subsequence ((model sequence-model) sequence)
+  "Add a record for all sub sequences. E.g. transform '(3 2 1)' into:
```
Typo "subsequences"

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+  "Add a record for all sub sequences. E.g. transform '(3 2 1)' into:
+'(3 2 1), '(2 1), '(1)"
+  (let ((sequence (copy-list sequence)))
+    (loop while (> (length sequence) 1)
```
Because it's `>` and not `>=` it seems that you are not adding the sequence of length 1.

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+(defmethod add-record-subsequence ((model sequence-model) sequence)
+  "Add a record for all sub sequences. E.g. transform '(3 2 1)' into:
+'(3 2 1), '(2 1), '(1)"
+  (let ((sequence (copy-list sequence)))
```
You don't have to `copy-list` if instead of `pop` you just assign the `rest` of the list of `sequence`.

jmercouris replied:
True!

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+
+(defmethod predict-subsequence-simple ((model sequence-model) sequence)
+  "Predict a sequence's next value based on all subsequence predictions. This is
+a naieve implementation which simply considers the amount of occurences without
```
Typo: "naive"

jmercouris replied:
I see. I was translating from German where the umlaut would make `ï->ie`

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+(defmethod predict-subsequence-simple ((model sequence-model) sequence)
+  "Predict a sequence's next value based on all subsequence predictions. This is
+a naieve implementation which simply considers the amount of occurences without
+regard to the weight of different chain lengths."
```
So what about the weighted version?  Shouldn't be too hard to work on.  Any blocker?

jmercouris replied:
No blocker!

Ambrevar commented on libraries/analysis/composite-sequence.lisp:
```diff
@@ -0,0 +1,88 @@
+  "Predict a sequence's next value based on all subsequence predictions. This is
+a naieve implementation which simply considers the amount of occurences without
+regard to the weight of different chain lengths."
+  (let* ((sequence (copy-list sequence))
```
Same here, no need for `copy-list` if you replace `pop`.

jmercouris replied:
Done!

Ambrevar commented on source/command-commands.lisp:
```diff
@@ -45,6 +45,19 @@ While disabled-mode commands are not listed, it's still possible to call them
+      (declare (ignore source))
+      (list (analysis:element
+             (analysis:predict (command-model *browser*)
+                               (list (last-command *browser*))))))))
```
Does this properly handle the case when there is no last command or command model?

jmercouris replied:
No, and I thought about that, however I did not implement it since the prompt buffer already handles this.

Ambrevar reviewed (commented):
Nice!  I'll test this as soon as possible.

Ambrevar commented:
> * Bind the prediction to a keybinding
>       OR
> 
>     * Add a list of commands which are ignored by the model

I suggest both! :)

Ambrevar commented:
Tests seem to be stuck into an infinite loop.

jmercouris commented:
> Tests seem to be stuck into an infinite loop.

This is odd. The tests I've added pass instantly on my machine. Also, the tests did not seem to fail on the tests I added.

jmercouris commented:
> > * Bind the prediction to a keybinding
> >   OR
> >   
> >   * Add a list of commands which are ignored by the model
> 
> I suggest both! :)

Done!

jmercouris commented:
Ready for review two!

Ambrevar commented on source/input.lisp:
```diff
@@ -86,10 +86,16 @@ KEYCODE-LESS-DISPLAY (KEYCODE-DISPLAY)."
+                            nyxt/prompt-buffer-mode:select-previous
+                            nyxt/prompt-buffer-mode:select-next-source
+                            nyxt/prompt-buffer-mode:select-previous-source)))
+    (unless (find (name (symbol-function function)) ignored-commands :test #'equal)
```
No need for `:test #'equal` here.

aartaka commented:
Note that there's alex:lastcar for exactly this idiom :)

aadcg commented:
What's the status here?

jmercouris commented:
Status is ready to merge.

jmercouris commented:
I shall merge it, let me resolve changelog issues, and then do so.
